src/Stratis.Features.FederatedPeg/TargetChain/InputConsolidator.cs

```diff
@@ -0,0 +1,250 @@
+        {
+            lock (this.lockObj)
+            {
+                // No need to sign in these cases.
+                if (!this.signingInProgress || this.fullySigned)
+                    return ConsolidationSignatureResult.Failed();
+
+                // Attempt to merge signatures
+                var builder = new TransactionBuilder(this.network);
+                Transaction oldTransaction = this.PartialTransaction;
+
+                this.logger.LogDebug("Attempting to merge signatures for {0} and {1}.", this.PartialTransaction.GetHash(), incomingPartialTransaction.GetHash());
+
+                this.PartialTransaction = SigningUtils.CombineSignatures(builder, this.PartialTransaction, new []{incomingPartialTransaction});
+
+                if (oldTransaction.GetHash() == this.PartialTransaction.GetHash())
+                {
+                    // Signing didn't work if the hash is still the same
+                    this.logger.LogDebug("Signing failed.");
+                    return ConsolidationSignatureResult.Failed();
+                }
+
+                this.logger.LogDebug("Successfully signed transaction.");
+
+                // NOTE: We don't need to reserve the transaction. The wallet will be at a standstill whilst this is happening.
+
+                // If it is FullySigned, broadcast.
+                if (this.walletManager.ValidateConsolidatingTransaction(this.PartialTransaction, true))
+                {
+                    this.logger.LogDebug("Consolidation transaction is fully signed. Broadcasting {0}", this.PartialTransaction.GetHash());
+                    this.broadcasterManager.BroadcastTransactionAsync(this.PartialTransaction);
+                    this.fullySigned = true;
+                }
+
+                return ConsolidationSignatureResult.Succeeded(this.PartialTransaction);
+            }
+        }
+
+        /// <summary>
+        /// Build a consolidating transaction.
+        /// </summary>
+        public Transaction BuildConsolidatingTransaction()
+        {
+            try
+            {
+                // TODO: Confifm that we can remove the ordering below.
+
+                List<UnspentOutputReference> unspentOutputs = this.walletManager.GetSpendableTransactionsInWallet(WithdrawalTransactionBuilder.MinConfirmations).ToList();
+
+                if (unspentOutputs.Count < WithdrawalTransactionBuilder.MaxInputs)
+                    throw new Exception("We shouldn't be consolidating transactions if we have less than 50 UTXOs to spend.");
+
+                IEnumerable<UnspentOutputReference> orderedUnspentOutputs = DeterministicCoinOrdering.GetOrderedUnspentOutputs(unspentOutputs);
+                IEnumerable<UnspentOutputReference> selectedInputs = orderedUnspentOutputs.Take(WithdrawalTransactionBuilder.MaxInputs);
+                string walletPassword = this.walletManager.Secret.WalletPassword;
+                bool sign = (walletPassword ?? "") != "";
```

Comment:
So if the wallet password is empty we don't sign?

Reply:
That's right... I'm actually not sure what this code is for, it's from `WithdrawalTransactionBuilder`. I figured maybe it stopped exceptions in cases when the wallet was still starting up but now that seems silly? Can we mark this to be checked in a future PR?

Reply:
OK yeah future PR :)